### PR TITLE
Support remote ci

### PR DESCRIPTION
**Pull request by SparkSnail**
Add integration test for remoteTrainingService platform

### CODE REVIEW

File: test/remote_docker.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    '''Start docker container'''
+    port = find_port()
+    source_dir = '/tmp/nnitest/' + name
+    run_cmds = ['docker', 'run', '-d', '-p', str(port) + ':22', '--name', name, '--mount', 'type=bind,source=' + source_dir + ',target=/tmp/nni', image]
+    output = check_output(run_cmds)
+    commit_id = output.decode('utf-8')
+    exec_cmds = ['docker', 'exec', name, 'python3', '-m', 'pip', 'install', '--user', '--no-cache-dir', '/tmp/nni/pynni/']
+    check_call(exec_cmds)
+    with open(source_dir + '/port', 'w') as file:
+        file.write('    port: ' + str(port))
```
Comment (chicm-ms) [inline test/remote_docker.py]: Suggest write only the port value to file,  then format the string 'port:  xxx' from outside.

File: azure-pipelines-daily.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+  steps:
+  - script: python3 -m pip install --upgrade pip setuptools
+    displayName: 'Install python tools'
+  - script: |
+      source install.sh
+    displayName: 'Install nni toolkit via source code'
+  - script: |
+      cd test
+      PATH=$HOME/Library/Python/3.7/bin:$PATH && source unittest.sh
```
Comment (xuehui1991) [inline azure-pipelines-daily.yml]: why there is "3.7"....

File: test/naive_test.py [hunk 3]
```diff
@@ -32,23 +34,25 @@
...
+    proc = subprocess.run(['nnictl', 'create', '--config', os.path.join('naive_test', config_file)])
     assert proc.returncode == 0, '`nnictl create` failed with code %d' % proc.returncode
 
     print('Spawning trials...')
 
     nnimanager_log_path = fetch_nni_log_path(EXPERIMENT_URL)
     current_trial = 0
 
-    for _ in range(120):
+    for _ in range(180):
```
Comment (xuehui1991) [inline test/naive_test.py]: magic number?

File: test/config_test.py [hunk 4]
```diff
@@ -39,8 +39,10 @@ def gen_new_config(config_file, training_service='local'):
     new_config_file = config_file + '.tmp'
 
     ts = get_yml_content('training_service.yml')[training_service]
-    print(config)
-    print(ts)
+    if training_service == 'remote':
+        ts['machineList'][0]['port'] = int(os.environ['docker_port'])
+        ts['machineList'][0]['username'] = os.environ['docker_user']
```
Comment (chicm-ms) [inline test/config_test.py]: Please do not use environment variables here,  write remote machine configuration into training_service.yml,  then call this config_test.py

File: test/training_service.yml [hunk 5]
```diff
@@ -4,7 +4,7 @@ local:
 remote:
   trainingServicePlatform: remote
   machineList:
-    - ip:
+    - ip: 23.96.57.139
```
Comment (chicm-ms) [inline test/training_service.yml]: This value should be written  at run time.

File: test/naive_test/remote.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+authorName: nni
```
Comment (chicm-ms) [inline test/naive_test/remote.yml]: This remote.yml should be removed.

File: test/sdk_test/remote.yml [hunk 7]
```diff
@@ -0,0 +1,24 @@
+assessor:
```
Comment (chicm-ms) [inline test/sdk_test/remote.yml]: Please remove this remote.yml

File: test/naive_test.py [hunk 8]
```diff
@@ -23,27 +23,31 @@
 import sys
 import time
 import traceback
+import os
+import argparse
 
 from utils import is_experiment_done, fetch_nni_log_path, read_last_line, remove_files, setup_experiment
 from utils import GREEN, RED, CLEAR, EXPERIMENT_URL
```
Comment (chicm-ms) [inline test/naive_test.py]: naive_test.py do not need to support remote machine

File: test/config_test/examples/cifar10-pytorch.test.yml [hunk 9]
```diff
@@ -3,6 +3,7 @@ experimentName: default_test
 maxExecDuration: 15m
 maxTrialNum: 2
 trialConcurrency: 2
+nniManagerIp: 52.168.146.89
```
Comment (chicm-ms) [inline test/config_test/examples/cifar10-pytorch.test.yml]: Please do not write nniManagerIp in *.test.yml,  write it in training_service.yml

File: test/config_test/examples/mnist-annotation.test.yml [hunk 10]
```diff
@@ -3,6 +3,7 @@ experimentName: default_test
 maxExecDuration: 5m
 maxTrialNum: 2
 trialConcurrency: 1
+nniManagerIp: 52.168.146.89
```
Comment (chicm-ms) [inline test/config_test/examples/mnist-annotation.test.yml]: Please do not write nniManagerIp in *.test.yml,  write it in training_service.yml

File: test/config_test/examples/mnist-keras.test.yml [hunk 11]
```diff
@@ -4,6 +4,7 @@ maxExecDuration: 5m
 maxTrialNum: 4
 trialConcurrency: 2
 searchSpacePath: ../../../examples/trials/mnist-keras/search_space.json
+nniManagerIp: 52.168.146.89
```
Comment (chicm-ms) [inline test/config_test/examples/mnist-keras.test.yml]: ditto

File: test/config_test/multi_phase/multi_phase.test.yml [hunk 12]
```diff
@@ -1,9 +1,10 @@
 authorName: nni
 experimentName: default_test
-maxExecDuration: 5m
+maxExecDuration: 10m
```
Comment (chicm-ms) [inline test/config_test/multi_phase/multi_phase.test.yml]: ditto

File: test/training_service.yml (hunk 5 above)
Comment (SparkSnail) [reply test/training_service.yml]: removed.

File: test/naive_test/remote.yml (hunk 6 above)
Comment (SparkSnail) [reply test/naive_test/remote.yml]: removed.

File: test/sdk_test/remote.yml (hunk 7 above)
Comment (SparkSnail) [reply test/sdk_test/remote.yml]: removed.

File: test/naive_test.py (hunk 8 above)
Comment (SparkSnail) [reply test/naive_test.py]: removed remote mode in naive_test.py

File: test/config_test/examples/cifar10-pytorch.test.yml (hunk 9 above)
Comment (SparkSnail) [reply test/config_test/examples/cifar10-pytorch.test.yml]: remvoed from example.yml

File: test/config_test/examples/mnist-annotation.test.yml (hunk 10 above)
Comment (SparkSnail) [reply test/config_test/examples/mnist-annotation.test.yml]: moved to training_service.yml

File: test/config_test/examples/mnist-keras.test.yml (hunk 11 above)
Comment (SparkSnail) [reply test/config_test/examples/mnist-keras.test.yml]: removed.

File: test/config_test/multi_phase/multi_phase.test.yml (hunk 12 above)
Comment (SparkSnail) [reply test/config_test/multi_phase/multi_phase.test.yml]: changed to 5m.

File: test/config_test.py [hunk 13]
```diff
@@ -130,4 +130,4 @@ def run(args):
 
     setup_experiment(args.preinstall)
 
-    run(args)
+    run(args)
```
Comment (chicm-ms) [inline test/config_test.py]: Unnecessary change.

File: pipelines-it-remote.yml [hunk 14]
```diff
@@ -0,0 +1,64 @@
+jobs:
+- job: 'Ubuntu_16_04_remote'
```
Comment (chicm-ms) [inline pipelines-it-remote.yml]: Please use a meaningful name to describe the job

File: pipelines-it-remote.yml [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+      scp -i test/id_rsa $(remote_user)@$(remote_host):/tmp/nnitest/$(Build.BuildId)/port test/port
+      cat test/port
+    displayName: 'Get docker port'
+  - script: |
+      cd test
+      python3 generate_ts_config.py --ts remote --remote_user $(remote_user) --remote_host $(remote_host) \
+      --remote_port $(cat port) --remote_pwd $(remote_pwd) --nni_manager_ip $(nni_manager_ip)
+      cat training_service.yml
+      PATH=$HOME/.local/bin:$PATH python3 config_test.py --ts remote --exclude multi_phase,multi_thread,cifar10
+    displayName: 'Built-in dispatcher tests'
```
Comment (chicm-ms) [inline pipelines-it-remote.yml]: Task name is incorrect.

File: pipelines-it-remote.yml [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+      sshEndpoint: remote_nni-ci-gpu-01
+      runOptions: commands
+      commands: python3 /tmp/nnitest/$(Build.BuildId)/test/remote_docker.py --mode start --name $(Build.BuildId) --image nni/nni
+    displayName: 'Start docker'
+  - task: DownloadSecureFile@1
+    inputs:
+      secureFile: remote_ci_private_key
+  - script: |
+      mkdir ~/.ssh
+      ssh-keyscan $(remote_host) >> ~/.ssh/known_hosts
```
Comment (chicm-ms) [inline pipelines-it-remote.yml]: Will this add one line into know_hosts for each build?

File: test/remote_docker.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+def find_port():
+    '''Find a port which is free'''
+    port = random.randint(10000, 20000)
+    while detect_port(port):
+        port = random.randint(10000, 20000)
+    return port
+
+def start_container(image, name):
+    '''Start docker container'''
```
Comment (chicm-ms) [inline test/remote_docker.py]: Please comment where the port file is stored.

File: test/pipelines-it-remote.yml [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+      chmod 600 test/id_rsa
+      scp -i test/id_rsa $(remote_user)@$(remote_host):/tmp/nnitest/$(Build.BuildId)/port test/port
+      cat test/port
+    displayName: 'Get docker port'
+  - script: |
+      cd test
+      python3 generate_ts_config.py --ts remote --remote_user $(remote_user) --remote_host $(remote_host) \
+      --remote_port $(cat port) --remote_pwd $(remote_pwd) --nni_manager_ip $(nni_manager_ip)
+      cat training_service.yml
+      PATH=$HOME/.local/bin:$PATH python3 config_test.py --ts remote --exclude multi_phase,multi_thread,cifar10
```
Comment (chicm-ms) [inline test/pipelines-it-remote.yml]: why exclude cifar10 and multi_thread?

File: test/config_test.py (hunk 13 above)
Comment (SparkSnail) [reply test/config_test.py]: recovered.

File: pipelines-it-remote.yml (hunk 14 above)
Comment (SparkSnail) [reply pipelines-it-remote.yml]: changed to 'integration_test_remote'

File: pipelines-it-remote.yml (hunk 15 above)
Comment (SparkSnail) [reply pipelines-it-remote.yml]: changed to 'integration test'

File: test/remote_docker.py (hunk 17 above)
Comment (SparkSnail) [reply test/remote_docker.py]: added comment.

File: test/pipelines-it-remote.yml [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    displayName: 'integration test'
+  - task: SSH@0
+    inputs:
+      sshEndpoint: remote_nni-ci-gpu-01
+      runOptions: commands
+      commands: python3 /tmp/nnitest/$(Build.BuildId)/test/remote_docker.py --mode stop --name $(Build.BuildId)
+    displayName: 'Stop docker'
+
+
+
```
Comment (chicm-ms) [inline test/pipelines-it-remote.yml]: It seems no new line at end of file.

File: pipelines-it-remote.yml (hunk 16 above)
Comment (SparkSnail) [reply pipelines-it-remote.yml]: removed

File: test/pipelines-it-remote.yml (hunk 18 above)
Comment (SparkSnail) [reply test/pipelines-it-remote.yml]: add this two examples

File: test/pipelines-it-remote.yml (hunk 19 above)
Comment (SparkSnail) [reply test/pipelines-it-remote.yml]: removed.

File: test/pipelines-it-remote.yml [hunk 20]
```diff
@@ -0,0 +1,59 @@
+jobs:
+- job: 'integration_test_remote'
+  pool: 'NNI CI KUBE CLI'
```
Comment (yds05) [inline test/pipelines-it-remote.yml]: The remote CI's agent pool is shared with KUBE?

Comment (yds05) [inline test/pipelines-it-remote.yml]: Remoe pool in our yml file. Suggest to configure it in Specific Azure pipeline.

Comment (SparkSnail) [reply test/pipelines-it-remote.yml]: removed in yaml file.

Comment (SparkSnail) [reply test/pipelines-it-remote.yml]: removed in yml file.
